### DISCUSSION TITLE
Memory management during copy

### DISCUSSION THREAD

**Discussion by malachib** [nlohmann/json#4257] [Q&A] [answered]
Firstly, let me say thank you for this awesome library @nlohmann !

OK.  Question: When utilizing the copy (not move) assignment operator, what copy technique is employed?  I can think of 3 possibilities:

- Deep copy (a complete reallocation of all keys, including subkeys)
- Shallow copy (just the one pointer)
- shared_ptr style - where there's a hidden ref counter at play

Any of these works for me, but I don't know which are at play!  Thank you

Related:  When I try to use the copy constructor in an initializer list, it tosses me into an array rabbit hole regardless of the `const json& copy_from` underlying  type.  Feels like a bug

**Comment by gregmarr** [comment, marked as answer]
The single assignment operator uses the by-value-parameter-and-swap idiom.  It takes the RHS `json` object by value, and then swaps contents with that by-value parameter.

If the assignment operator is passed an rvalue, then the by-value parameter is constructed using the move constructor, otherwise it's constructed using the copy constructor.  The copy constructor does a deep copy.  The move constructor just steals the value.

The initializer list constructor can be a bit weird at times, and the behavior is actually different depending on which compiler you are using.  https://github.com/nlohmann/json/issues/2311

**Reply by malachib** [reply]
If I understand you correctly, the single assignment operator behaves a lot like a move operation.

In a scenario of two `json` instances where `a` is empty and `b` has content, am I to understand `a=b` means `a` will have a shallow (pointer) copy of `b` and `b` will now be empty?

**Reply by gregmarr** [reply]
>If I understand you correctly, the single assignment operator behaves a lot like a move operation.

Only when given an rvalue.

> In a scenario of two json instances where a is empty and b has content, am I to understand a=b means a will have a shallow (pointer) copy of b and b will now be empty?

No, `a = b;` will be a copy.  `a = std::move(b);` will behave as you described, and the same for `a = makeJson()` where `makeJson()` returns `json`.

**Reply by malachib** [reply]
That makes sense, makeJson() return would be an rvalue.  As long as RVO is at play, I might be able to avoid some fuss.
